packages/tangle/tipmanager.go

```diff
@@ -324,9 +327,207 @@ func (t *TipManager) Tips(p payload.Payload, countParents int) (parents MessageI
+		})
+	})
+	return true
+}
+
+func (t *TipManager) processMarker(pastMarker *markers.Marker, minSupportedTimestamp time.Time, oldestUnconfirmedMarker *markers.Marker) (tscValid bool) {
+	// oldest unconfirmed marker is in the future cone of the past marker (same sequence), therefore past marker is confirmed and there is no need to check
+	if pastMarker.Index() < oldestUnconfirmedMarker.Index() {
+		return true
+	}
+
+	return t.getMarkerMessage(oldestUnconfirmedMarker).IssuingTime().After(minSupportedTimestamp)
+}
+
+func (t *TipManager) checkMessage(messageID MessageID, messageWalker *walker.Walker, minSupportedTimestamp time.Time) (timestampValid bool) {
+	timestampValid = true
+
+	if t.tangle.ConfirmationOracle.IsMessageConfirmed(messageID) {
+		return
+	}
+	t.tangle.Storage.Message(messageID).Consume(func(message *Message) {
+		if message.IssuingTime().Before(minSupportedTimestamp) {
+			timestampValid = false
+			messageWalker.StopWalk()
+			return
+		}
+
+		// walk through strong parents' past cones
+		for parentID := range message.ParentsByType(StrongParentType) {
+			messageWalker.Push(parentID)
+		}
+	})
+	return timestampValid
+}
+
+func (t *TipManager) getMarkerMessage(marker *markers.Marker) (markerMessage *Message) {
+	messageID := t.tangle.Booker.MarkersManager.MessageID(marker)
+	if messageID == EmptyMessageID {
+		panic(fmt.Errorf("failed to retrieve marker message for %s", marker.String()))
+	}
+	t.tangle.Storage.Message(messageID).Consume(func(message *Message) {
+		markerMessage = message
+	})
 	return
 }
 
+// getOldestUnconfirmedMarker is similar to FirstUnconfirmedMarkerIndex, except it skips any marker gaps an existing marker.
+func (t *TipManager) getOldestUnconfirmedMarker(pastMarker *markers.Marker) *markers.Marker {
+	unconfirmedMarkerIdx := t.tangle.ConfirmationOracle.FirstUnconfirmedMarkerIndex(pastMarker.SequenceID())
+
+	// skip any gaps in marker indices
+	for ; unconfirmedMarkerIdx <= pastMarker.Index(); unconfirmedMarkerIdx++ {
+		currentMarker := markers.NewMarker(pastMarker.SequenceID(), unconfirmedMarkerIdx)
+
+		// Skip if there is no marker at the given index, i.e., the sequence has a gap.
+		if t.tangle.Booker.MarkersManager.MessageID(currentMarker) == EmptyMessageID {
+			continue
+		}
+		break
+	}
+
+	oldestUnconfirmedMarker := markers.NewMarker(pastMarker.SequenceID(), unconfirmedMarkerIdx)
+	return oldestUnconfirmedMarker
+}
+
+func (t *TipManager) getPreviousConfirmedIndex(sequence *markers.Sequence, markerIndex markers.Index) markers.Index {
+	// skip any gaps in marker indices
+	for ; sequence.LowestIndex() <= markerIndex; markerIndex-- {
+		currentMarker := markers.NewMarker(sequence.ID(), markerIndex)
+
+		// Skip if there is no marker at the given index, i.e., the sequence has a gap or marker is not yet confirmed (should not be the case).
+		if t.tangle.Booker.MarkersManager.MessageID(currentMarker) == EmptyMessageID || !t.tangle.ConfirmationOracle.IsMarkerConfirmed(currentMarker) {
```

Comment:
Same here: `IsMarkerConfirmed` internally has to do another round of `MarkersManager.MessageID(currentMarker)`. So better to use `IsMessageConfirmed` directly.

Reply:
:+1: